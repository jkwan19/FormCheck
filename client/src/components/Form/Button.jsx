import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Button from '@material-ui/core/Button';
import CloudUploadIcon from '@material-ui/icons/CloudUpload';
import SaveIcon from '@material-ui/icons/Save';


const useStyles = makeStyles((theme) => ({
  root: {
    '& > *': {
      margin: theme.spacing(1),
    },
  },
  input: {
    display: 'none',
  },
}));

function UploadButtons(props) {
  const classes = useStyles();

  return (
    <div className={classes.root}>
      <input
        accept="image/*"
        className={classes.input}
        id="contained-button-file"
        multiple
        type="file"
        accept="image/png, image/jpeg"
        onChange={props.uploadImage}
      />
      <div>
        <label htmlFor="contained-button-file">
          <Button
          variant="contained"
          color="default"
          component="span"
          className={classes.button}
          startIcon={<CloudUploadIcon />}
        >
          Upload
          </Button>
        </label>
        {props.image ? props.image.name : ''}
      </div>
      <Button
        type="submit"
        onClick={props.upload}
        variant="contained"
        color="primary"
        size="medium"
        className={classes.button}
        startIcon={<SaveIcon />}
      >
        Save
      </Button>
    </div>
  );
}

export default UploadButtons;