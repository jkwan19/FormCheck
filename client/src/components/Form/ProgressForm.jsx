import React, { useState, useEffect } from 'react';
import styled, { createGlobalStyle } from 'styled-components';
import axios from 'axios';
import UploadButtons from './Button.jsx';

const GlobalStyles = createGlobalStyle`
  @import url('https://fonts.googleapis.com/css?family=Notable');
  body {
    font-family: 'Notable', sans-serif;
  }`;

const Wrapper = styled('div')``;
const Form = styled('form')``;

const Input = styled('input')`
  width: 100%;
  padding: 12px 20px;
  margin: 8px 0;
  display: inline-block;
  border: 1px solid #ccc;
  border-radius: 4px;
  box-sizing: border-box;
`;
const Options = styled('option')``;
const Dropdown = styled('select')`
  width: 100%;
  padding: 12px 20px;
  margin: 8px 0;
  display: inline-block;
  border: 1px solid #ccc;
  border-radius: 4px;
  box-sizing: border-box;
`;

function ProgressForm(props) {
  const [workout, setWorkout] = useState('Shoulders');
  const [image, setImage] = useState('')

  const workoutHandler = (e) => {
    setWorkout(e.target.value);
  };

  const imageHandler = (e) => {
    setImage(e.target.files[0]);
  };

  const uploadHandler = (e) => {
    e.preventDefault();
    const formData = new FormData();
    formData.append('image', image);
    let data = {
      workout,
      image: formData
    }
    props.handleForm(data);
    setWorkout('Shoulders');
    setImage('');
  };

  return (
    <Wrapper>
      <GlobalStyles />
      <Form method="post" enctype="application/x-www-form-urlencoded" action="./public/uploads">
        Workout:
        <Dropdown onChange={workoutHandler}>
          <Options value="Shoulders">Shoulders</Options>
          <Options value="Planks">Planks</Options>
        </Dropdown>
        Upload Image
        <UploadButtons upload={uploadHandler} uploadImage={imageHandler} image={image}/>
      </Form>
    </Wrapper>
  );
}

export default ProgressForm;
