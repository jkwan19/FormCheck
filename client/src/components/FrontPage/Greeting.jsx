import React, { useState, useEffect } from 'react';
import styled from 'styled-components';
import Typography from '@material-ui/core/Typography';

const Wrapper = styled("div")`
`;
const Header = styled("h4")`
`;
const Body = styled("p")`
`;

const Greeting = () => (
  <Wrapper>
    <Typography paragraph variant="h4">
      Welcome to Form Check!
    </Typography>
    <Typography paragraph>
      Form Check is a great way to count your reps and executing them to perfection, while avoiding injuries.
    </Typography>
  </Wrapper>
)

export default Greeting;