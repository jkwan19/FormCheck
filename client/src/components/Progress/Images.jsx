import React, { useState, useEffect } from 'react';
import styled from 'styled-components';
import moment from 'moment';
import { makeStyles } from '@material-ui/core/styles';
import GridList from '@material-ui/core/GridList';
import GridListTile from '@material-ui/core/GridListTile';
import GridListTileBar from '@material-ui/core/GridListTileBar';
import IconButton from '@material-ui/core/IconButton';
import StarBorderIcon from '@material-ui/icons/StarBorder';


const useStyles = makeStyles((theme) => ({
  root: {
    display: 'flex',
    flexWrap: 'wrap',
    justifyContent: 'space-around',
    overflow: 'hidden',
    float: 'center',
    backgroundColor: theme.palette.background.paper,
  },
  gridList: {
    width: 'auto',
    height: 'auto',
    // Promote the list into his own layer on Chrome. This cost memory but helps keeping high FPS.
    transform: 'translateZ(0)',
  },
  titleBar: {
    background:
      'linear-gradient(to bottom, rgba(0,0,0,0.7) 0%, ' +
      'rgba(0,0,0,0.3) 70%, rgba(0,0,0,0) 100%)',
  },
  icon: {
    color: 'white',
  },
}));

function Images(props) {
  const classes = useStyles();
  const progressList = props.progress;
  const toDate = (date) => {
    return moment(date).calendar();
  }
  console.log(progressList, 'list')
  if (progressList.length > 0) {
    return (
      <div className={classes.root}>
        <GridList cellHeight={200} spacing={1} className={classes.gridList}>
          {progressList.map((tile, index) => (
            <GridListTile key={`${tile.image}${index}`} cols={index === 0 ? 2 : 1} rows={index === 0 ? 2 : 1}>
              <img src={tile.image} alt={tile.workout} />
              <GridListTileBar
                title={tile.workout}
                subtitle={toDate(tile.createdAt)}
                titlePosition="top"
                actionIcon={
                  <IconButton aria-label={`star ${tile.workout}`} className={classes.icon}>
                  </IconButton>
                }
                actionPosition="left"
                className={classes.titleBar}
              />
            </GridListTile>
          ))}
        </GridList>
      </div>
    );
  } else {
    return <div>{'Loading.....'}</div>
  }
}

export default Images;
