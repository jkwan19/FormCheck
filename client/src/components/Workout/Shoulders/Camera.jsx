import React, { useRef, useReducer, useCallback, useState, useEffect } from "react"

function getKeypointsObject(pose) {
  return pose.keypoints.reduce((acc, { part, position }) => {
    acc[part] = position
    return acc
  }, {})
}
function reducer(count, action) {
  if (action === "reset") {
    return 0;
  }
  return count + 1;
}
let isGoodForm;
function Camera (sensitivity = 10) {
  const [count, dispatch] = useReducer(reducer, 0);
  const standard = useRef(0);
  let isComingDown = false;
  const goingUp = () => {
    isComingDown = false;
  }
  const goingDown = () => {
    isComingDown = true;
  }
  const checkPoses = useCallback(
    poses => {
      if (poses.length !== 1) {
        return;
      }
      const {
        leftShoulder,
        rightShoulder,
        leftElbow,
        rightElbow,
        leftWrist,
        rightWrist
      } = getKeypointsObject(poses[0]);
      const elbow = leftElbow || rightElbow;
      const shoulder = leftShoulder || rightShoulder;
      if (!elbow || !shoulder) {
        return;
      }
      const wrist = leftWrist || rightWrist;
      const up = shoulder.y > elbow.y;
      if (wrist && up) {
        goingDown();
        standard.current = Math.max(standard.current, elbow.y);
        return;
      }
      const down = standard.current < (elbow.y + sensitivity);
      const downRange = (bodyPosition) => {
        return (300 <= Math.floor(bodyPosition) <= 360);
      };
      const tooLow = (bodyPosition) => {
        return (bodyPosition > 340);
      }
      if ((down && isComingDown) && !!wrist) {
        console.log('down');
        goingUp();
        dispatch("increment");
        //shoulder and elbow need to be in range of 300 => 320
        if (downRange(shoulder.y) && downRange(elbow.y)) {
          isGoodForm = true;
        }
        standard.current = 0
        return;
      }
      if (tooLow(elbow.y) && !isComingDown && !!wrist) {
        isGoodForm = false;
      }
      const rest = !wrist;
      if (rest) {
        dispatch("reset");
      }
    },
    [sensitivity]
  );
  return [count, checkPoses];
}
export {Camera, isGoodForm};
