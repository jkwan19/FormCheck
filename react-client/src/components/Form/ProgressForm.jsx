import React, { useState, useEffect } from 'react';
import styled, {createGlobalStyle} from 'styled-components';
import axios from 'axios';

const GlobalStyles = createGlobalStyle`
  @import url('https://fonts.googleapis.com/css?family=Notable');
  body {
    font-family: 'Notable', sans-serif;
  }`

const Wrapper = styled('div')``;
const Input = styled('input')`
  width: 100%;
  padding: 12px 20px;
  margin: 8px 0;
  display: inline-block;
  border: 1px solid #ccc;
  border-radius: 4px;
  box-sizing: border-box;
`;
const Dropdown = styled('select')`
  width: 100%;
  padding: 12px 20px;
  margin: 8px 0;
  display: inline-block;
  border: 1px solid #ccc;
  border-radius: 4px;
  box-sizing: border-box;
`;

function ProgressForm (props) {
  // const [selectedFile, setSelectedFile] = useState(null);
  const [title, setTitle] = useState('Shoulders');
  const [imageUrl, setImageUrl] = useState('');

  const titleHandler = (e) => {
    setTitle(e.target.value);
  }

  const imageHandler = (e) => {
    setImageUrl(e.target.value);
  }

  const uploadHandler = (e) => {
    e.preventDefault();
    let formData = {
      workout: title,
      imageUrl: imageUrl
    }
    props.handleForm(formData);
    setTitle('Shoulders');
    setImageUrl('');
  }

  return (
    <Wrapper>
      <GlobalStyles />
      <form onSubmit={uploadHandler}>
        Workout:
        <Dropdown onChange={titleHandler}>
          <option value="Shoulders">Shoulders</option>
          <option value="Planks">Planks</option>
        </Dropdown>
        Image: <Input type="text" onChange={imageHandler} placeholder="Link..."/>
        <button type="submit">Upload!</button>
      </form>
    </Wrapper>
    )
}

export default ProgressForm;