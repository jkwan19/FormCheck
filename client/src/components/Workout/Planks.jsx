import React, { useState, useEffect } from 'react';
import styled from 'styled-components';
import YouTube from 'react-youtube';
import CheckForm from './Planks/CheckForm.jsx';
import { makeStyles } from '@material-ui/core/styles';
import Button from '@material-ui/core/Button';

const Wrapper = styled("div")``;
const Header = styled("h2")`
  font: 1.1 em;
  text-align: center;
`;

const VideoContainer = styled("div")`
  text-align: center;
  margin: auto;
`

function Planks (props) {
  const [ form, setForm ] = useState(false);
  const [ workout, setWorkout ] = useState("Planks");
  const opts = {
    height: '300',
    width: '510',
    playerVars: {
      // https://developers.google.com/youtube/player_parameters
      autoplay: 1,
    },
  }
  const onReady = (e) => {
    // access to player in all event handlers via event.target
    e.target.pauseVideo();
  }
  const handleClick = (e) => {
    e.preventDefault();
    setForm(true);
  }
  const renderView = () => {
    if (form === true) {
      return <CheckForm workout={workout}/>
    } else {
      return (
        <Wrapper>
          <Header>Planks Demo</Header>
          <VideoContainer>
            <YouTube videoId="pSHjTRCQxIw" opts={opts} onReady={onReady}/>
          </VideoContainer>
          <Button onClick={handleClick} variant="contained" color="primary">
            Check My Form
          </Button>
        </Wrapper>
      )
    }
  }
  return (
    <Wrapper>
      {renderView()}
    </Wrapper>
  )
}

export default Planks