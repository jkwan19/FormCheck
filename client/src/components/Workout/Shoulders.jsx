import React, { useState, useEffect } from 'react';
import styled from 'styled-components';
import YouTube from 'react-youtube';
import CheckForm from './Shoulders/CheckForm.jsx';
import { makeStyles } from '@material-ui/core/styles';
import Button from '@material-ui/core/Button';


const Wrapper = styled("div")`
  text-align: center;
`;
const Header = styled("h2")`
  font: 1.1 em;
`;

const VideoContainer = styled("div")`
  text-align: center;
  margin: auto;
`;
function Shoulders (props) {
  const [ form, setForm ] = useState(false);
  const [ workout, setWorkout ] = useState("Shoulder Press");
  const opts = {
    height: '300',
    width: '510',
    playerVars: {
      // https://developers.google.com/youtube/player_parameters
      autoplay: 1,
    },
  }

  const onReady = (e) => {
    // access to player in all event handlers via event.target
    e.target.pauseVideo();
  }

  const handleClick = (e) => {
    e.preventDefault();
    setForm(true);
  }

  const renderView = () => {
    if (form === true) {
      return <CheckForm workout={workout}/>
    } else {
      return (
        <Wrapper>
          <Header>Dumbbell Shoulder Press Demo</Header>
          <VideoContainer>
            <YouTube videoId="B-aVuyhvLHU" opts={opts} onReady={onReady}/>
          </VideoContainer>
          <Button onClick={handleClick} variant="contained" color="primary">
            Check My Form
          </Button>
        </Wrapper>
      )
    }
  }
  return (
    <Wrapper>
      {renderView()}
    </Wrapper>
  )
}
export default Shoulders;