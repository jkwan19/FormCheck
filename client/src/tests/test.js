import React from 'react';
import axios from 'axios';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';

// Components
import App from '../components/FrontPage/App.jsx';
import Greeting from '../components/FrontPage/Greeting.jsx';
import SearchBar from '../components/FrontPage/SearchBar.jsx';
import Images from '../components/Progress/Images.jsx';
import Planks from '../components/Workout/Planks.jsx';
import Shoulders from '../components/Workout/Shoulders.jsx';
import Button from '../components/Form/Button.jsx';
import ProgressForm from '../components/Form/ProgressForm.jsx';

jest.mock('axios');

// App Component
describe('App', () => {
  test('should render App', () => {
    const wrapper = shallow(
      <App />
    );
    expect(wrapper.exists()).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
  })
  test('should test Apps component with default state of empty array', () => {
    const wrapper = shallow(<App workouts={[]} />);
    expect(wrapper).toMatchSnapshot();
   });

  test("should test good response", () => {
    axios.get.mockImplementation(() => Promise.resolve({ data: {} }));
  });

  test("should test bad response", () => {
    axios.get.mockImplementation(() => Promise.reject({ }));
  });

  it("displays initial progresses", () => {
    const wrapper = mount(<App />);
    expect(wrapper.find("images")).toHaveLength(0);
  });
})

// Greeting Component
describe('Greeting', () => {
  test('should render Greeting component', () => {
    const wrapper = shallow(
      <Greeting />
    );
    expect(wrapper.exists()).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
  })
})

//SearchBar Component
describe('SearchBar', () => {
  test('should render SearchBar component', () => {
    const wrapper = shallow(
      <SearchBar />
    );
    expect(wrapper.exists()).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
  })
})

//Images Component
describe('Images', () => {
  test("Renders correctly", () => {
    const props = {
      options: [{ key: 'key',text: 'text',value: 'value'}],
      value: 'value',
      placeholder: 'placeholder',
      onSelect: jest.fn(),
      children: <div>test</div>
    }
    const wrapper = shallow(<Images {...props}/>);
    expect(wrapper.exists()).toBe(true);
  });
})

//Planks Component
describe('Planks', () => {
  test('should render Planks component', () => {
    const wrapper = shallow(
      <Planks />
    );
    expect(wrapper.exists()).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
  })
})

//Shoulders Component
describe('Shoulders', () => {
  test('should render Shoulders component', () => {
    const wrapper = shallow(
      <Shoulders />
    );
    expect(wrapper.exists()).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
  })
})

//Form Button Component
describe('Form Button', () => {
  test('should render form Button component', () => {
    const wrapper = shallow(
      <Button />
    );
    expect(wrapper.exists()).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
  })
  it('should disable submit button on submit click', () => {
    const wrapper = mount(<Button />);
    const submitButton = wrapper.find(Button);
    submitButton.simulate('click');

    setTimeout(() => {
      expect(submitButton.prop('disabled')).toBeTruthy();
    });
  });

})

//Progress Form Component
describe('Progress Form', () => {
  test('should render Progress Form component', () => {
    const wrapper = shallow(
      <ProgressForm />
    );
    expect(wrapper.exists()).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
  })

  it('Changing select button should call setState', () => {
    const onChange = jest.fn();
    const wrapper = mount(<ProgressForm onChange={onChange} value='Shoulders'/>);
    wrapper.find('input').simulate('change');
    expect(onChange).toBeCalledWith('Shoulders');
  });
})
