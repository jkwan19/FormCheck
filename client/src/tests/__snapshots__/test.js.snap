// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render 1`] = `ShallowWrapper {}`;

exports[`App should render App 1`] = `ShallowWrapper {}`;

exports[`App should test Apps component with default state of empty array 1`] = `ShallowWrapper {}`;

exports[`Form Button should render form Button component 1`] = `ShallowWrapper {}`;

exports[`Greeting should render Greeting 1`] = `ShallowWrapper {}`;

exports[`Greeting should render Greeting component 1`] = `ShallowWrapper {}`;

exports[`Planks should render Planks component 1`] = `ShallowWrapper {}`;

exports[`Progress Form should handle onSubmit handler 1`] = `ShallowWrapper {}`;

exports[`Progress Form should render Progress Form component 1`] = `ShallowWrapper {}`;

exports[`SearchBar should render SearchBar component 1`] = `ShallowWrapper {}`;

exports[`Shoulders should render Shoulders component 1`] = `ShallowWrapper {}`;
